File: frameworks/web-components/entities/ChessRules.js
const OPPOSITE = {
  white: "black",
  black: "white"
};

const BACK_RANK = ["rook", "knight", "bishop", "queen", "king", "bishop", "knight", "rook"];

function clonePiece(piece) {
  return piece ? { ...piece } : piece;
}

function normalizeType(piece) {
  return String(piece?.type || piece?.kind || "").toLowerCase();
}

function normalizeColor(piece) {
  return String(piece?.color || "").toLowerCase();
}

function isEnemy(piece, color) {
  return piece && normalizeColor(piece) !== color;
}

function addSlidingMoves(board, piece, row, col, directions, moves) {
  const color = normalizeColor(piece);
  for (const direction of directions) {
    let nextRow = row + direction.row;
    let nextCol = col + direction.col;

    while (board.inBounds(nextRow, nextCol)) {
      const target = board.get(nextRow, nextCol);
      if (!target) {
        moves.push({ from: { row, col }, to: { row: nextRow, col: nextCol }, capture: null });
      } else {
        if (isEnemy(target, color)) {
          moves.push({
            from: { row, col },
            to: { row: nextRow, col: nextCol },
            capture: { row: nextRow, col: nextCol, piece: target }
          });
        }
        break;
      }

      nextRow += direction.row;
      nextCol += direction.col;
    }
  }
}

export class ChessRules {
  constructor(options = {}) {
    this.autoPromoteTo = options.autoPromoteTo || "queen";
  }

  setupStandardBoard(board) {
    for (let row = 0; row < 8; row += 1) {
      for (let col = 0; col < 8; col += 1) {
        board.set(row, col, null);
      }
    }

    for (let col = 0; col < 8; col += 1) {
      board.set(0, col, { type: BACK_RANK[col], color: "black" });
      board.set(1, col, { type: "pawn", color: "black" });
      board.set(6, col, { type: "pawn", color: "white" });
      board.set(7, col, { type: BACK_RANK[col], color: "white" });
    }

    return board;
  }

  legalMoves(board, row, col, options = {}) {
    const piece = board.get(row, col);
    if (!piece) {
      return [];
    }

    const pseudoMoves = this.pseudoLegalMoves(board, row, col);
    if (options.allowSelfCheck === true) {
      return pseudoMoves;
    }

    const color = normalizeColor(piece);
    return pseudoMoves.filter((move) => !this._wouldLeaveKingInCheck(board, move, color));
  }

  pseudoLegalMoves(board, row, col) {
    const piece = board.get(row, col);
    if (!piece) {
      return [];
    }

    const type = normalizeType(piece);
    const color = normalizeColor(piece);
    const moves = [];

    if (type === "pawn") {
      this._pawnMoves(board, piece, row, col, moves);
    } else if (type === "knight") {
      this._jumpMoves(board, piece, row, col, [
        [-2, -1], [-2, 1], [-1, -2], [-1, 2],
        [1, -2], [1, 2], [2, -1], [2, 1]
      ], moves);
    } else if (type === "bishop") {
      addSlidingMoves(board, piece, row, col, [
        { row: -1, col: -1 }, { row: -1, col: 1 },
        { row: 1, col: -1 }, { row: 1, col: 1 }
      ], moves);
    } else if (type === "rook") {
      addSlidingMoves(board, piece, row, col, [
        { row: -1, col: 0 }, { row: 1, col: 0 },
        { row: 0, col: -1 }, { row: 0, col: 1 }
      ], moves);
    } else if (type === "queen") {
      addSlidingMoves(board, piece, row, col, [
        { row: -1, col: -1 }, { row: -1, col: 1 },
        { row: 1, col: -1 }, { row: 1, col: 1 },
        { row: -1, col: 0 }, { row: 1, col: 0 },
        { row: 0, col: -1 }, { row: 0, col: 1 }
      ], moves);
    } else if (type === "king") {
      this._jumpMoves(board, piece, row, col, [
        [-1, -1], [-1, 0], [-1, 1],
        [0, -1], [0, 1],
        [1, -1], [1, 0], [1, 1]
      ], moves);
      this._castlingMoves(board, piece, row, col, moves);
    }

    return moves.map((move) => ({
      ...move,
      piece: clonePiece(piece),
      color
    }));
  }

  applyMove(board, move) {
    const piece = board.get(move.from.row, move.from.col);
    if (!piece) {
      return false;
    }

    const nextPiece = clonePiece(piece);
    const type = normalizeType(nextPiece);
    if (type === "pawn" && (move.to.row === 0 || move.to.row === 7)) {
      nextPiece.type = move.promotion || this.autoPromoteTo;
    }

    board.set(move.from.row, move.from.col, null);
    board.set(move.to.row, move.to.col, { ...nextPiece, moved: true });

    // Castling — a king move flagged with `castle` also slides the rook. The UI passes
    // the full move object (from legalMoves) straight to applyMove, so the `castle`
    // payload survives and both pieces move together.
    if (move.castle && move.castle.rook) {
      const rookFrom = move.castle.rook.from;
      const rookTo = move.castle.rook.to;
      const rook = board.get(rookFrom.row, rookFrom.col);
      if (rook) {
        board.set(rookFrom.row, rookFrom.col, null);
        board.set(rookTo.row, rookTo.col, { ...clonePiece(rook), moved: true });
      }
    }
    return true;
  }

  allLegalMoves(board, color) {
    const moves = [];
    board.forEach((piece, row, col) => {
      if (piece && normalizeColor(piece) === color) {
        moves.push(...this.legalMoves(board, row, col));
      }
    });
    return moves;
  }

  isKingInCheck(board, color) {
    const king = this._findKing(board, color);
    if (!king) {
      return false;
    }
    return this.isSquareAttacked(board, king.row, king.col, OPPOSITE[color]);
  }

  isCheckmate(board, color) {
    return this.isKingInCheck(board, color) && this.allLegalMoves(board, color).length === 0;
  }

  isStalemate(board, color) {
    return !this.isKingInCheck(board, color) && this.allLegalMoves(board, color).length === 0;
  }

  isSquareAttacked(board, row, col, byColor) {
    let attacked = false;
    board.forEach((piece, pieceRow, pieceCol) => {
      if (attacked || !piece || normalizeColor(piece) !== byColor) {
        return;
      }

      const attacks = this._attackSquares(board, pieceRow, pieceCol);
      attacked = attacks.some((square) => square.row === row && square.col === col);
    });
    return attacked;
  }

  _pawnMoves(board, piece, row, col, moves) {
    const color = normalizeColor(piece);
    const direction = color === "white" ? -1 : 1;
    const startRow = color === "white" ? 6 : 1;
    const oneStep = { row: row + direction, col };
    const twoStep = { row: row + direction * 2, col };

    if (board.inBounds(oneStep.row, oneStep.col) && !board.get(oneStep.row, oneStep.col)) {
      moves.push({ from: { row, col }, to: oneStep, capture: null });
      if (row === startRow && !board.get(twoStep.row, twoStep.col)) {
        moves.push({ from: { row, col }, to: twoStep, capture: null });
      }
    }

    for (const captureCol of [col - 1, col + 1]) {
      if (!board.inBounds(row + direction, captureCol)) {
        continue;
      }
      const target = board.get(row + direction, captureCol);
      if (isEnemy(target, color)) {
        moves.push({
          from: { row, col },
          to: { row: row + direction, col: captureCol },
          capture: { row: row + direction, col: captureCol, piece: target }
        });
      }
    }
  }

  _jumpMoves(board, piece, row, col, offsets, moves) {
    const color = normalizeColor(piece);
    for (const [dr, dc] of offsets) {
      const targetRow = row + dr;
      const targetCol = col + dc;
      if (!board.inBounds(targetRow, targetCol)) {
        continue;
      }

      const target = board.get(targetRow, targetCol);
      if (!target || isEnemy(target, color)) {
        moves.push({
          from: { row, col },
          to: { row: targetRow, col: targetCol },
          capture: target ? { row: targetRow, col: targetCol, piece: target } : null
        });
      }
    }
  }

  _attackSquares(board, row, col) {
    const piece = board.get(row, col);
    const type = normalizeType(piece);
    const color = normalizeColor(piece);

    if (type === "pawn") {
      const direction = color === "white" ? -1 : 1;
      return [
        { row: row + direction, col: col - 1 },
        { row: row + direction, col: col + 1 }
      ].filter((square) => board.inBounds(square.row, square.col));
    }

    if (type === "king") {
      // A king attacks the 8 adjacent squares. Compute them directly (NOT via
      // pseudoLegalMoves) — otherwise _castlingMoves → isSquareAttacked → _attackSquares
      // would recurse infinitely, and castling squares would count as "attacked".
      return [
        [-1, -1], [-1, 0], [-1, 1],
        [0, -1], [0, 1],
        [1, -1], [1, 0], [1, 1]
      ].map(([dr, dc]) => ({ row: row + dr, col: col + dc }))
       .filter((square) => board.inBounds(square.row, square.col));
    }

    return this.pseudoLegalMoves(board, row, col).map((move) => move.to);
  }

  // Castling (2026-07-01). King slides two squares toward an unmoved rook when: neither
  // has moved, the squares between are empty, the king is not in check, and it does not
  // pass through (or land on) an attacked square. Emits a move carrying `castle` so
  // applyMove also slides the rook. `moved` is stamped by applyMove; fresh pieces have
  // it unset, so `!piece.moved` == hasn't moved.
  _castlingMoves(board, piece, row, col, moves) {
    if (piece.moved) return;
    const color = normalizeColor(piece);
    const homeRow = color === "white" ? 7 : 0;
    if (row !== homeRow || col !== 4) return;         // king must be on its start square
    const enemy = OPPOSITE[color];
    if (this.isSquareAttacked(board, row, col, enemy)) return;   // can't castle out of check

    // Kingside: rook on col 7; f,g (5,6) empty; king passes f,g unattacked.
    const kRook = board.get(row, 7);
    if (kRook && normalizeType(kRook) === "rook" && normalizeColor(kRook) === color && !kRook.moved
        && !board.get(row, 5) && !board.get(row, 6)
        && !this.isSquareAttacked(board, row, 5, enemy)
        && !this.isSquareAttacked(board, row, 6, enemy)) {
      moves.push({
        from: { row, col }, to: { row, col: 6 }, capture: null,
        castle: { side: "kingside", rook: { from: { row, col: 7 }, to: { row, col: 5 } } }
      });
    }

    // Queenside: rook on col 0; b,c,d (1,2,3) empty; king passes d,c (3,2) unattacked.
    const qRook = board.get(row, 0);
    if (qRook && normalizeType(qRook) === "rook" && normalizeColor(qRook) === color && !qRook.moved
        && !board.get(row, 1) && !board.get(row, 2) && !board.get(row, 3)
        && !this.isSquareAttacked(board, row, 3, enemy)
        && !this.isSquareAttacked(board, row, 2, enemy)) {
      moves.push({
        from: { row, col }, to: { row, col: 2 }, capture: null,
        castle: { side: "queenside", rook: { from: { row, col: 0 }, to: { row, col: 3 } } }
      });
    }
  }

  _findKing(board, color) {
    let king = null;
    board.forEach((piece, row, col) => {
      if (!king && piece && normalizeColor(piece) === color && normalizeType(piece) === "king") {
        king = { row, col, piece };
      }
    });
    return king;
  }

  _wouldLeaveKingInCheck(board, move, color) {
    const piece = board.get(move.from.row, move.from.col);
    const captured = board.get(move.to.row, move.to.col);

    board.set(move.from.row, move.from.col, null);
    board.set(move.to.row, move.to.col, piece);
    const inCheck = this.isKingInCheck(board, color);
    board.set(move.from.row, move.from.col, piece);
    board.set(move.to.row, move.to.col, captured);

    return inCheck;
  }
}

File: frameworks/macos-components/components/board-game/ChessRules.swift
// BEGIN mChatAI macOS Component: boardgame.chess (components/board-game/ChessRules.swift)
import Foundation

enum ChessRules {
    static func startingBoard() -> Board8x8 {
        var board = Board8x8()
        let backRank: [BoardPieceKind] = [.rook, .knight, .bishop, .queen, .king, .bishop, .knight, .rook]
        for col in 0..<8 {
            board[PuzzlePoint(row: 0, col: col)] = BoardPiece(side: .black, kind: backRank[col])
            board[PuzzlePoint(row: 1, col: col)] = BoardPiece(side: .black, kind: .pawn)
            board[PuzzlePoint(row: 6, col: col)] = BoardPiece(side: .white, kind: .pawn)
            board[PuzzlePoint(row: 7, col: col)] = BoardPiece(side: .white, kind: backRank[col])
        }
        return board
    }

    static func legalMoves(from point: PuzzlePoint, on board: Board8x8, filterCheck: Bool = true) -> [BoardMove] {
        guard let piece = board[point] else { return [] }
        let pseudoMoves: [BoardMove]

        switch piece.kind {
        case .pawn:
            pseudoMoves = pawnMoves(from: point, piece: piece, board: board)
        case .rook:
            pseudoMoves = rayMoves(from: point, piece: piece, board: board, directions: orthogonalDirections)
        case .bishop:
            pseudoMoves = rayMoves(from: point, piece: piece, board: board, directions: diagonalDirections)
        case .queen:
            pseudoMoves = rayMoves(from: point, piece: piece, board: board, directions: orthogonalDirections + diagonalDirections)
        case .king:
            pseudoMoves = stepMoves(from: point, piece: piece, board: board, directions: orthogonalDirections + diagonalDirections)
        case .knight:
            pseudoMoves = stepMoves(from: point, piece: piece, board: board, directions: knightDirections)
        case .checker, .checkerKing:
            pseudoMoves = []
        }

        guard filterCheck else { return pseudoMoves }
        return pseudoMoves.filter { move in
            var next = board
            apply(move, to: &next)
            return !isKingInCheck(piece.side, on: next)
        }
    }

    static func allLegalMoves(for side: BoardSide, on board: Board8x8) -> [BoardMove] {
        board.pieces(for: side).flatMap { legalMoves(from: $0.point, on: board) }
    }

    static func apply(_ move: BoardMove, to board: inout Board8x8) {
        guard var piece = board[move.from] else { return }
        board[move.from] = nil
        if let promotion = move.promotion {
            piece.kind = promotion
        }
        board[move.to] = piece
    }

    static func isKingInCheck(_ side: BoardSide, on board: Board8x8) -> Bool {
        guard let king = board.pieces(for: side).first(where: { $0.piece.kind == .king })?.point else {
            return false
        }
        for enemy in board.pieces(for: side.opponent) {
            if legalMoves(from: enemy.point, on: board, filterCheck: false).contains(where: { $0.to == king }) {
                return true
            }
        }
        return false
    }

    static func isCheckmate(_ side: BoardSide, on board: Board8x8) -> Bool {
        isKingInCheck(side, on: board) && allLegalMoves(for: side, on: board).isEmpty
    }

    private static let orthogonalDirections = [
        PuzzlePoint(row: -1, col: 0), PuzzlePoint(row: 1, col: 0),
        PuzzlePoint(row: 0, col: -1), PuzzlePoint(row: 0, col: 1)
    ]

    private static let diagonalDirections = [
        PuzzlePoint(row: -1, col: -1), PuzzlePoint(row: -1, col: 1),
        PuzzlePoint(row: 1, col: -1), PuzzlePoint(row: 1, col: 1)
    ]

    private static let knightDirections = [
        PuzzlePoint(row: -2, col: -1), PuzzlePoint(row: -2, col: 1),
        PuzzlePoint(row: -1, col: -2), PuzzlePoint(row: -1, col: 2),
        PuzzlePoint(row: 1, col: -2), PuzzlePoint(row: 1, col: 2),
        PuzzlePoint(row: 2, col: -1), PuzzlePoint(row: 2, col: 1)
    ]

    private static func pawnMoves(from point: PuzzlePoint, piece: BoardPiece, board: Board8x8) -> [BoardMove] {
        let step = piece.side == .white ? -1 : 1
        let startRow = piece.side == .white ? 6 : 1
        let promotionRow = piece.side == .white ? 0 : 7
        var moves: [BoardMove] = []

        let one = PuzzlePoint(row: point.row + step, col: point.col)
        if board.contains(one), board[one] == nil {
            moves.append(BoardMove(from: point, to: one, promotion: one.row == promotionRow ? .queen : nil))
            let two = PuzzlePoint(row: point.row + step * 2, col: point.col)
            if point.row == startRow, board.contains(two), board[two] == nil {
                moves.append(BoardMove(from: point, to: two))
            }
        }

        for colDelta in [-1, 1] {
            let target = PuzzlePoint(row: point.row + step, col: point.col + colDelta)
            if board.contains(target), let occupant = board[target], occupant.side != piece.side {
                moves.append(BoardMove(from: point, to: target, captured: target, promotion: target.row == promotionRow ? .queen : nil))
            }
        }

        return moves
    }

    private static func stepMoves(from point: PuzzlePoint, piece: BoardPiece, board: Board8x8, directions: [PuzzlePoint]) -> [BoardMove] {
        directions.compactMap { direction in
            let target = PuzzlePoint(row: point.row + direction.row, col: point.col + direction.col)
            guard board.contains(target) else { return nil }
            if let occupant = board[target], occupant.side == piece.side { return nil }
            return BoardMove(from: point, to: target, captured: board[target] == nil ? nil : target)
        }
    }

    private static func rayMoves(from point: PuzzlePoint, piece: BoardPiece, board: Board8x8, directions: [PuzzlePoint]) -> [BoardMove] {
        var moves: [BoardMove] = []
        for direction in directions {
            var target = PuzzlePoint(row: point.row + direction.row, col: point.col + direction.col)
            while board.contains(target) {
                if let occupant = board[target] {
                    if occupant.side != piece.side {
                        moves.append(BoardMove(from: point, to: target, captured: target))
                    }
                    break
                }
                moves.append(BoardMove(from: point, to: target))
                target = PuzzlePoint(row: target.row + direction.row, col: target.col + direction.col)
            }
        }
        return moves
    }
}
// END mChatAI macOS Component: boardgame.chess
